File: cr5.zig
const std = @import("std");
const math = std.math;

fn letterValue(c: u8) i32 {
    return @intCast(i32, c - 'a' + 1);
}

fn letterSumArray(s: []const u8) void {
    var first = true;
    std.debug.print("[", .{});
    for (s) |c| {
        if ((c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z')) {
            if (!first) std.debug.print(", '+', ", .{});
            std.debug.print("{d}", .{letterValue(c)});
            first = false;
        }
    }
    std.debug.print("]\n", .{});
}

fn partitions(n: i32, arr: []i32, pos: usize) void {
    if (n == 0) {
        std.debug.print("[", .{});
        for (arr[0..pos]) |v, i| {
            if (i != 0) std.debug.print(", '+', ", .{});
            std.debug.print("{d}", .{v});
        }
        std.debug.print("]\n", .{});
        return;
    }
    for (i32(1)..=n) |i| {
        arr[pos] = i;
        partitions(n-i, arr, pos+1);
    }
}

fn specialFormula(n: i32) void {
    std.debug.print("[2, '*', {d}, '-', {d}]\n", .{n, n});
}

fn meaning(sum: i32) []const u8 {
    return switch(sum){
        1 => "BOX",
        6 => "ZAHL",
        7 => "WINKEL",
        9 => "RAUM",
        else => "ANDERE"
    };
}

fn printWinkel(a: i32, b: i32) void {
    std.debug.print("[1, '+', {d}, '+', {d}]\n", .{a,b});
}

fn printRaum(dim_count: i32, angles: []f64, points: [][]f64, ids: [][]const u8) void {
    std.debug.print("// RAUM: dim_count={d}\n", .{dim_count});
    for (angles) |angle, i| {
        std.debug.print("Angle {d}: {f}\n", .{i+1, angle});
        std.debug.print("Point {d}: ({f},{f},{f}) id={s}\n", .{i+1, points[i][0], points[i][1], points[i][2], ids[i]});
    }
}

pub fn main() void {
    const cmd = "abc";

    std.debug.print("// Buchstabenaufsummierung:\n", .{});
    letterSumArray(cmd);

    std.debug.print("// Alle Aufsummierungen von 3:\n", .{});
    var arr: [10]i32 = undefined;
    partitions(3, arr[0..], 0);

    std.debug.print("// Spezielle Formel für 9:\n", .{});
    specialFormula(9);

    std.debug.print("// Bedeutungen:\n", .{});
    for (1..=9) |s| {
        std.debug.print("Summe {d}: {s}\n", .{s, meaning(s)});
    }

    std.debug.print("// Beispiel WINKEL (Summe 7):\n", .{});
    printWinkel(2,3);

    std.debug.print("// Beispiel RAUM (Summe 9):\n", .{});
    const dim_count: i32 = 3;
    const angles: [3]f64 = [3]f64{0.5, math.pi/4, math.pi/6};
    const points: [[3]f64] = [[3]f64{0,0,0}, [3]f64{1,0,0}, [3]f64{0,1,0}];
    const ids: [_][]const u8 = [_][]const u8{"p1","p2","p3"};
    printRaum(dim_count, angles[0..], points[0..], ids);
}

File: cr5.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

int letter_value(char c) {
    return tolower(c) - 'a' + 1;
}

// Buchstaben aufsummieren + '+'
void letter_sum_array(const char *s) {
    printf("[");
    int first = 1;
    for(int i=0; s[i]; i++) {
        if(isalpha(s[i])) {
            if(!first) printf(", '+', ");
            printf("%d", letter_value(s[i]));
            first = 0;
        }
    }
    printf("]\n");
}

// Alle möglichen Aufsummierungen von n
void partitions(int n, int *arr, int pos) {
    if(n==0) {
        printf("[");
        for(int i=0;i<pos;i++){
            if(i>0) printf(", '+', ");
            printf("%d", arr[i]);
        }
        printf("]\n");
        return;
    }
    for(int i=1;i<=n;i++){
        arr[pos] = i;
        partitions(n-i, arr, pos+1);
    }
}

// Spezielle Formel n=((2*n)-n) mit '*' und '-'
void special_formula(int n) {
    printf("[2, '*', %d, '-', %d]\n", n, n);
}

// Bedeutung nach Summe
const char* meaning(int sum) {
    switch(sum){
        case 1: return "BOX";
        case 6: return "ZAHL";
        case 7: return "WINKEL";
        case 9: return "RAUM";
        default: return "ANDERE";
    }
}

// Winkel → 2 Zahlen ohne Betrag + 1
void print_winkel(int a, int b) {
    printf("[1, '+', %d, '+', %d]\n", a, b);
}

// Raum → Dimensionen, Winkel, Punkte, IDs
void print_raum(int dim_count, double angles[], double points[][3], const char* ids[]) {
    printf("// RAUM: dim_count=%d\n", dim_count);
    for(int i=0;i<dim_count;i++){
        printf("Angle %d: %f\n", i+1, angles[i]);
        printf("Point %d: (%f,%f,%f) id=%s\n", i+1, points[i][0], points[i][1], points[i][2], ids[i]);
    }
}

int main() {
    const char *cmd = "abc";

    printf("// Buchstabenaufsummierung:\n");
    letter_sum_array(cmd);

    printf("// Alle Aufsummierungen von 3:\n");
    int arr[10];
    partitions(3, arr, 0);

    printf("// Spezielle Formel für 9:\n");
    special_formula(9);

    printf("// Bedeutungen:\n");
    for(int s=1;s<=9;s++){
        printf("Summe %d: %s\n", s, meaning(s));
    }

    printf("// Beispiel WINKEL (Summe 7):\n");
    print_winkel(2,3);

    printf("// Beispiel RAUM (Summe 9):\n");
    int dim_count = 3;
    double angles[3] = {0.5, M_PI/4, M_PI/6};
    double points[3][3] = {{0,0,0},{1,0,0},{0,1,0}};
    const char* ids[3] = {"p1","p2","p3"};
    print_raum(dim_count, angles, points, ids);
}
